src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -191,6 +200,37 @@ void writeGenerators()
                 }
                 writer.WriteArrayEnd();
             }
+
+            void writeResources()
+            {
+                writer.WriteArrayStart();
+                foreach (ResourceDescription resource in resources)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    writer.WriteObjectStart();
+                    writer.Write("resourceName", resource.ResourceName);
+                    writer.Write("fileName", resource.FileName);
+                    writer.Write("isPublic", resource.IsPublic);
+                    writer.WriteKey("content");
+                    writeResourceContent(writer, resource);
+                    writer.WriteObjectEnd();
+                }
+                writer.WriteArrayEnd();
+            }
+
+            void writeResourceContent(JsonWriter writer, ResourceDescription resource)
+            {
+                if (resource.IsEmbedded)
+                {
+                    using var stream = resource.DataProvider();
+                    var sourceText = SourceText.From(stream, encoding: null, checksumAlgorithm: SourceHashAlgorithms.Default);
+                    WriteSourceText(writer, sourceText);
```

Comment:
Why are you converting to a `SourceText` here? These resources are generally treated as binary blobs in the codebase. My intuition would be that we just grab the raw `Stream` or `byte[]` and hash that. Not sure if we'd hit weird edge cases trying to view a binary as a source text.